app/data_client.go

```diff
@@ -761,3 +848,355 @@ func (d *DataClient) RemoveBinaryDataFromDatasetByIDs(
+	}
+	return response, nil
+}
+
+// DataCaptureUpload uploads the metadata and contents for either tabular or binary data,
+// and returns the file ID associated with the uploaded data and metadata.
+func (d *DataClient) DataCaptureUpload(ctx context.Context, metadata UploadMetadata, sensorContents []SensorData) (string, error) {
+	resp, err := d.dataSyncClient.DataCaptureUpload(ctx, &syncPb.DataCaptureUploadRequest{
+		Metadata:       uploadMetadataToProto(metadata),
+		SensorContents: sensorContentsToProto(sensorContents),
+	})
+	if err != nil {
+		return "", err
+	}
+	return resp.FileId, nil
+}
+
+// StreamingDataCaptureUpload uploads metadata and streaming binary data in chunks.
+func (d *DataClient) StreamingDataCaptureUpload(
+	ctx context.Context,
+	data []byte,
+	partID string,
+	fileExt string,
+	options *StreamingOptions,
+) (string, error) {
+	uploadMetadata := UploadMetadata{
+		PartID:           partID,
+		ComponentType:    options.ComponentType,
+		ComponentName:    options.ComponentName,
+		MethodName:       options.MethodName,
+		Type:             DataTypeBinarySensor,
+		FileName:         options.FileName,
+		MethodParameters: options.MethodParameters,
+		FileExtension:    fileExt,
+		Tags:             options.Tags,
+	}
+	uploadMetadataPb := uploadMetadataToProto(uploadMetadata)
+	var sensorMetadata SensorMetadata
+	if len(options.DataRequestTimes) == 2 {
+		sensorMetadata = SensorMetadata{
+			TimeRequested: options.DataRequestTimes[0],
+			TimeReceived:  options.DataRequestTimes[1],
+		}
+	}
+	sensorMetadataPb := sensorMetadataToProto(sensorMetadata)
+	metadata := &syncPb.DataCaptureUploadMetadata{
+		UploadMetadata: uploadMetadataPb,
+		SensorMetadata: sensorMetadataPb,
+	}
+	// establish a streaming connection.
+	stream, err := d.dataSyncClient.StreamingDataCaptureUpload(ctx)
+	if err != nil {
+		return "", err
+	}
+	// send the metadata as the first packet.
+	metaReq := &syncPb.StreamingDataCaptureUploadRequest{
+		UploadPacket: &syncPb.StreamingDataCaptureUploadRequest_Metadata{
+			Metadata: metadata,
+		},
+	}
+	if err := stream.Send(metaReq); err != nil {
+		return "", err
+	}
+
+	// send the binary data in chunks.
+	for start := 0; start < len(data); start += UploadChunkSize {
+		end := start + UploadChunkSize
+		if end > len(data) {
+			end = len(data)
+		}
+		dataReq := &syncPb.StreamingDataCaptureUploadRequest{
+			UploadPacket: &syncPb.StreamingDataCaptureUploadRequest_Data{
+				Data: data[start:end],
+			},
+		}
+		if err := stream.Send(dataReq); err != nil {
+			return "", err
+		}
+	}
+	// close the stream and get the response.
+	resp, err := stream.CloseAndRecv()
+	if err != nil {
+		return "", err
+	}
+	return resp.FileId, nil
+}
+
+// FileUploadFromBytes uploads the contents and metadata for binary data such as encoded images or other data represented by bytes.
+func (d *DataClient) FileUploadFromBytes(
+	ctx context.Context,
+	partID string,
+	data []byte,
+	opts *FileUploadOptions,
+) (string, error) {
```

Comment:
I see that this is an upload function, but what is it actually returning? It's not clear what this string value is supposed to represent.

Reply:
Ah, I see now from the private function you're calling it's the file ID, but it should be made clear in the comment.